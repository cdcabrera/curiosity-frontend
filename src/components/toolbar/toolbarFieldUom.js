import React from 'react';
import PropTypes from 'prop-types';
import { connect, reduxTypes, store } from '../../redux';
import { Select } from '../form/select';
import { RHSM_API_QUERY_UOM_TYPES as FIELD_TYPES, RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { translate } from '../i18n/i18n';

/**
 * Display a unit of measure (uom) field with options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {string} props.value
 * @param {Function} props.t
 * @param {string} props.viewId
 * @returns {Node}
 */
const ToolbarFieldUom = ({ value, t, viewId }) => {
  const options = Object.values(FIELD_TYPES).map(type => ({
    title: translate('curiosity-toolbar.uom', { context: type }),
    value: type,
    selected: type === value
  }));

  /**
   * On select, dispatch type.
   *
   * @event onSelect
   * @param {object} event
   * @returns {void}
   */
  const onSelect = (event = {}) =>
    store.dispatch([
      {
        type: reduxTypes.query.SET_QUERY_CLEAR_INVENTORY_LIST
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_TYPES[RHSM_API_QUERY_TYPES.UOM],
        viewId,
        [RHSM_API_QUERY_TYPES.UOM]: event.value
      }
    ]);

  return (
    <Select
      aria-label={t('curiosity-toolbar.placeholder', { context: 'uom' })}
      onSelect={onSelect}
      options={options}
      selectedOptions={value}
      placeholder={t('curiosity-toolbar.placeholder', { context: 'uom' })}
    />
  );
};

/**
 * Prop types.
 *
 * @type {{viewId: string, t: Function, value: string}}
 */
ToolbarFieldUom.propTypes = {
  t: PropTypes.func,
  value: PropTypes.oneOf([...Object.values(FIELD_TYPES)]),
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{viewId: string, t: translate, value: string}}
 */
ToolbarFieldUom.defaultProps = {
  t: translate,
  value: FIELD_TYPES.CORES,
  viewId: 'toolbarFieldUom'
};

/**
 * Apply state to props.
 *
 * @param {object} state
 * @param {object} state.view
 * @param {object} props
 * @param {string} props.viewId
 * @returns {object}
 */
const mapStateToProps = ({ view }, { viewId }) => ({
  value: view.query?.[RHSM_API_QUERY_TYPES.UOM]?.[viewId]
});

const ConnectedToolbarFieldUom = connect(mapStateToProps)(ToolbarFieldUom);

export { ConnectedToolbarFieldUom as default, ConnectedToolbarFieldUom, ToolbarFieldUom };
