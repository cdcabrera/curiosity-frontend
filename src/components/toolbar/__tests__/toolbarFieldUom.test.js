import React from 'react';
import { shallow } from 'enzyme';
import { ToolbarFieldUom } from '../toolbarFieldUom';

describe('ToolbarFieldUom Component', () => {
  it('should render a non-connected component', () => {
    const props = {};
    const component = shallow(<ToolbarFieldUom {...props} />);

    expect(component).toMatchSnapshot('non-connected');
  });
});
