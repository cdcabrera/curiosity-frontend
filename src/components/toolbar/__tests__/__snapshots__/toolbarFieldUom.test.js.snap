// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarFieldUom Component should render a non-connected component: non-connected 1`] = `
<Select
  aria-label="t(curiosity-toolbar.placeholder, {\\"context\\":\\"uom\\"})"
  ariaLabel="Select option"
  className=""
  id="generatedid-"
  isDisabled={false}
  isToggleText={true}
  name={null}
  onSelect={[Function]}
  options={
    Array [
      Object {
        "selected": true,
        "title": "t(curiosity-toolbar.uom, {\\"context\\":\\"cores\\"})",
        "value": "cores",
      },
      Object {
        "selected": false,
        "title": "t(curiosity-toolbar.uom, {\\"context\\":\\"sockets\\"})",
        "value": "sockets",
      },
    ]
  }
  placeholder="t(curiosity-toolbar.placeholder, {\\"context\\":\\"uom\\"})"
  selectedOptions="cores"
  toggleIcon={null}
  variant="single"
/>
`;
